cmd/release-controller/http.go

```diff
@@ -925,3 +1015,248 @@ var extendedRelTime = []humanize.RelTimeMagnitude{
+		var timeFormat, timeString, stream string
+		prevTags, _ := c.findReleaseStreamTags(false, latestPromotedFrom)
+		if prevTags[latestPromotedFrom] != nil &&
+				prevTags[latestPromotedFrom].Tag.Annotations[releaseAnnotationCreationTimestamp] != "" {
+			// Use previous release stream tags, if available
+			timeFormat = time.RFC3339
+			stream = prevTags[latestPromotedFrom].Tag.Annotations[releaseAnnotationName]
+			timeString = prevTags[latestPromotedFrom].Tag.Annotations[releaseAnnotationCreationTimestamp]
+		} else if rePreviousReleaseName.MatchString(latestPromotedFrom) {
+			// Try to use name format of previous nightly to find release stream and timestamp
+			timeFormat = releaseTimestampFormat
+			stream = rePreviousReleaseName.ReplaceAllString(latestPromotedFrom, "${STREAM}")
+			timeString = rePreviousReleaseName.ReplaceAllString(latestPromotedFrom, "${TIMESTAMP}")
+		} else {
+			glog.Errorf("Could not find tag %s , tag may have been deleted", latestPromotedFrom)
+			continue
+		}
+
+		// Check if selected stream belongs to any we are interested in
+		found = false
+		for _, s := range releaseStreams {
+			if stream == s {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// The stable release belongs to a release stream we are not processing
+			continue
+		}
+
+		pt, err := time.Parse(timeFormat, timeString)
+		if err != nil {
+			glog.Errorf("Unable to parse timestamp %s for %s: %v", timeString, latestPromotedFrom, err)
+			continue
+		}
+		remaining--
+		latestPromotedTime[stream] = pt.Unix()
+		next[stream] = &v
+	}
+
+	for _, stream := range releaseStreams {
+		nextReleaseName := ""
+		if next[stream] != nil {
+			nextVersion, err := incrementSemanticVersion(*next[stream])
+			if err == nil {
+				nextReleaseName = nextVersion.String()
+			} else {
+				glog.Errorf("Unable to increment semantic version %s", next[stream].String())
+			}
+		}
+		candidates := make([]*ReleaseCandidate, 0)
+		releaseTags := tagsForRelease(releaseStreamTagMap[stream].Release)
+		for _, tag := range releaseTags {
+			if tag.Annotations != nil && tag.Annotations[releaseAnnotationPhase] == releasePhaseAccepted &&
+					tag.Annotations[releaseAnnotationCreationTimestamp] != "" {
+				t, _ := time.Parse(time.RFC3339, tag.Annotations[releaseAnnotationCreationTimestamp])
+				ts := t.Unix()
+				if ts > latestPromotedTime[stream] {
+					candidates = append(candidates, &ReleaseCandidate{
+						ReleasePromoteJobParameters {
+							FromTag:tag.Name,
+							Name:nextReleaseName,
+						},
+						time.Unix(ts,0).Format(time.RFC3339),
+						tag,
+					})
+				}
+			}
+		}
+		sort.Slice(candidates, func(i, j int) bool {
+			return candidates[i].CreationTime > candidates[j].CreationTime
+		})
+		releaseCandidates[stream] = &ReleaseCandidateList{Items:candidates}
+	}
+	return releaseCandidates, nil
+}
+
+func (c *Controller) filterByUpgradeSuccess(rcList map[string]*ReleaseCandidateList, successPercent float64) map[string]*ReleaseCandidateList {
```

Comment:
Instead of making this a method on controller, make it a regular function and pass in the graph.  Generally we try to keep object methods focused on logical operations on the object (do X, do Y) and regular functions for the helpers for those methods, which only need a subset of what controller might offer.